ggshield/scan/repo.py

```diff
@@ -75,13 +77,56 @@ def scan_commit(
     except Exception as exc:
         results = Results.from_exception(exc)
 
-    return ScanCollection(
-        commit.sha or "unknown",
-        type="commit",
-        results=results,
-        optional_header=commit.optional_header,
-        extra_info=commit.info._asdict(),
-    )
+    scans = []
+    for commit in commits:
+        concerned_results = [
+            res
+            for res in results.results
+            if any(res.content == file.document for file in commit.files)
+        ]
+        scans.append(
+            ScanCollection(
+                commit.sha or "unknown",
+                type="commit",
+                results=Results(
+                    results=concerned_results,
+                    errors=results.errors,
+                ),
+                optional_header=commit.optional_header,
+                extra_info=commit.info._asdict(),
+            )
+        )
+
+    return ScanCollection(id=scan_context.command_id, type="commit-ranges", scans=scans)
+
+
+def get_commits_content_by_batch(
+    commit_list: List[str],
+    exclusion_regexes: Set[re.Pattern],
+    batch_max_size: int = MULTI_DOCUMENT_LIMIT,
+) -> Iterator[Tuple[Files, List[Commit]]]:
```

Comment:
The signature of this function is surprising: it returns the same information twice. Reading more of the code I guess this is because we need a single instance of `Files` to run `Files.scan()`. That's another justification for the refactoring in #344: if `Files.scan` were to become a `scan(List[File], ...)` function then this would not be a problem.

Until this is done, I think it would be better to hide that issue inside `scan_commits_content()`:

1. make `get_commits_content_by_batch()` return a `Iterator[List[Commit]]`
2. in `scan_commits_content()`, create a Files instance like this:

```python
commit_files = list(itertools.chain.from_iterable(c.files for c in commits))
files = Files(commit_files)
```

Reply:
The patch is only parsed once: `commit.files` caches the results, so this should not be a problem.